### PR TITLE
feat: add blog index page

### PR DESCRIPTION
**Pull request by smeijer**
## Added dependencies

- [`date-fns`](https://npmjs.com/date-fns) for date formatting

## Changes

I've:

- created the "blog index page" under path `/blog`
- added active/inactive states to the links in the navbar
- added a max-width to the grid, and thereby all content except navbar. The designs don't include this, but I really think it's better this way. Opening full-width pages on ultra-wide screens ain't nice.
- corrected the implementation of the "copy article url" button on the "article card" to show an icon on small and text on large screens.
- wrapped `ArrowButton` in a `button`, and added `ArrowLink` which wraps a `<Link>`
- connected the input and tag buttons to the url, not to the backend
- implemented the search state, that hides the featured article
- implemented the no-articles-found state, using a placeholder image as the 404 is absent.

## Todo's

- [x] show `n articles found` caption when filtering
- [x] implement "no articles found" state

### CODE REVIEW

File: app/components/tag.tsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import * as React from 'react'
+import clsx from 'clsx'
+
+interface TagProps {
+  tag: string
+  selected?: boolean
+}
+
+function Tag({tag, selected}: TagProps) {
+  // TODO: didn't know the behavior of these tags, and thereby if they should be a <button>, <Link> or @reach/checkbox
```
Comment (smeijer) [inline app/components/tag.tsx:10]: I didn't know the behavior of the tags/category buttons :point_up:

File: app/routes/blog/index.tsx [hunk 2]
```diff
@@ -33,22 +44,102 @@ export function meta() {
...
+              <SearchIcon />
+            </div>
+            <input
+              placeholder="Search blog"
+              aria-label="Search blog"
+              className="dark:focus:bg-gray-800 placeholder-black dark:placeholder-white px-16 py-6 w-full text-black dark:text-white text-lg font-medium focus:bg-gray-100 bg-transparent border border-gray-200 dark:border-gray-600 rounded-full focus:outline-none"
+            />
+            <div className="absolute right-8 top-0 flex items-center justify-center h-full text-blueGray-500 text-lg font-medium">
+              {/* TODO: replace count */}
+              173
```
Comment (smeijer) [inline app/routes/blog/index.tsx:87]: I've the number from the design, but I don't really know what it stands for. Is it the number of matching articles? Or the limit of characters that can be used as query?

File: app/routes/blog/index.tsx [hunk 3]
```diff
@@ -33,22 +44,102 @@ export function meta() {
...
+        </div>
+        <div className="flex flex-wrap col-span-full -ml-4 -mt-4 lg:col-span-10">
+          {/* TODO: get tags from database, and connect `selected` to something */}
+          {tags.map((tag, idx) => (
+            <Tag key={tag} tag={tag} selected={idx === 3} />
+          ))}
+        </div>
+      </Grid>
+
+      {/* TODO: this should be `smallest`, but `medium` looks better? */}
```
Comment (smeijer) [inline app/routes/blog/index.tsx:107]: Not sure if I made the right call here. I think that `medium` looks better. But please compare with `smallest` before merging.

File: app/routes/blog/index.tsx [hunk 4]
```diff
@@ -33,22 +44,102 @@ export function meta() {
...
+
+      <Spacer size="large" />
+      <div className="flex justify-center w-full">
+        <LoadMoreButton />
+      </div>
+      <Spacer size="large" />
+
+      <Grid>
+        <div className="col-span-full lg:col-span-5">
+          {/* TODO: replace image with canoe */}
```
Comment (smeijer) [inline app/routes/blog/index.tsx:130]: Can you upload the canoe image to `cloudinary`, so I can use it?

File: app/routes/index.tsx [hunk 5]
```diff
@@ -18,7 +20,8 @@ function IndexRoute() {
       <Spacer size="large" />
       <ProblemSolutionSection />
       <Spacer size="medium" />
-      <BlogSection articles={[]} />
+      {/*  TODO: replace fixtures */}
+      <BlogSection articles={articles as unknown as ArticleCardProps[]} />
```
Comment (smeijer) [inline app/routes/index.tsx:24]: I've added the fixtures here, so I could see if things were still looking okay.

**Comment by kentcdodds** [comment]
> The designs don't include this

Feel free to take creative liberties like this as needed :)

I just pushed an update to the images so all the illustrations should be there now (with alt text): cef1e60

File: app/components/load-more-button.tsx [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import * as React from 'react'
+
+function LoadMoreButton() {
+  return (
+    <button className="dark:focus:bg-gray-800 flex items-center px-8 py-6 dark:text-white focus:bg-gray-100 bg-transparent border border-gray-200 dark:border-gray-600 rounded-full focus:outline-none">
+      {/* TODO: an svg plus often looks better. */}
```
[the remark this answers is not in the document]
Comment (kentcdodds) [reply app/components/load-more-button.tsx:6]: Yeah, let's find a good SVG `+` 👍

File: app/routes/blog/index.tsx (hunk 2 above)
Comment (kentcdodds) [reply app/routes/blog/index.tsx:87]: I think it's based on what's currently on my blog: https://kentcdodds.com/blog

So it's the number of matching based on the filter.

**Comment by kentcdodds** [comment]
I rebased my changes for you 👍

File: app/routes/index.tsx (hunk 5 above)
Comment (kentcdodds) [reply app/routes/index.tsx:24]: That's fine. I'll work on this bit today.

File: app/components/tag.tsx [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+interface TagProps {
+  tag: string
+  selected?: boolean
+  onClick?: MouseEventHandler<HTMLButtonElement>
+}
+
+function Tag({tag, selected, onClick}: TagProps) {
+  // TODO: make this accessible, aria-checked and aria-selected aren't allowed on buttons
+  //   should we make it a checkbox with @reach/checkbox?
```
Comment (smeijer) [inline app/components/tag.tsx:13]: :point_up_2:

Comment (kentcdodds) [reply app/components/tag.tsx:13]: Yes, let's do that 👍

**Comment by kentcdodds** [comment]
I just pushed an update to wire up the blog page with the real posts. Search is not yet implemented, but the way the MDX files are loaded has been reworked quite a bit. I'll work on migrating more (all?) of my posts over to this repo soon so we have more real stuff to work with.

**Comment by kentcdodds** [comment]
I stayed up too late, but I got a lot done! https://github.com/kentcdodds/kentcdodds.com/commit/d7ebfcc7e3e8fb8354f0ca58da8159fccaeeb14d

I moved over a few of the posts. Hopefully that's enough to get you going on things. Now I'm going to sleep (it's almost 2 AM here).

File: app/components/tag.tsx (hunk 7 above)
Comment (smeijer) [reply app/components/tag.tsx:13]: done :sunglasses: 

tags

**Comment by smeijer** [comment]
All flags green. I'm moving to the next page.

**Comment by kentcdodds** [comment]
I guess I was too sleepy to realize that git hooks failed when I tried adding those posts 😅 They're added now. I'm going to work on automating copying all posts over to the repo now so we'll have all the posts on the site :)

**Pull request merged**